File: src/cc_proxy/project_store.py
"""Discover Claude Code projects from ~/.claude/projects/."""

import os
from pathlib import Path

from .models import ProjectInfo, SessionEntry
from .path_utils import decode_project_dir, encode_project_path, get_claude_projects_dir
from .session_store import is_alive, list_sessions


def list_projects(claude_home: str = "~/.claude") -> list[ProjectInfo]:
    projects_dir = get_claude_projects_dir(claude_home)
    if not os.path.isdir(projects_dir):
        return []

    sessions = list_sessions(claude_home)
    # Build dir_name -> cwd mapping from session data (authoritative)
    dir_to_cwd: dict[str, str] = {}
    cwd_to_sessions: dict[str, list[SessionEntry]] = {}
    for s in sessions:
        encoded = encode_project_path(s.cwd)
        dir_to_cwd[encoded] = s.cwd
        cwd_to_sessions.setdefault(s.cwd, []).append(s)

    projects = []
    for entry in sorted(os.listdir(projects_dir)):
        full_path = os.path.join(projects_dir, entry)
        if not os.path.isdir(full_path):
            continue

        real_path = dir_to_cwd.get(entry, decode_project_dir(entry))
        jsonl_count = len(list(Path(full_path).glob("*.jsonl")))
        project_sessions = cwd_to_sessions.get(real_path, [])
        has_active = any(is_alive(s.pid) for s in project_sessions)

        projects.append(ProjectInfo(
            path=real_path,
            encoded_name=entry,
            session_count=jsonl_count,
            has_active_session=has_active,
        ))

    return projects


def get_project_dir(project_path: str, claude_home: str = "~/.claude") -> str | None:
    encoded = encode_project_path(project_path)
    full_path = os.path.join(get_claude_projects_dir(claude_home), encoded)
    return full_path if os.path.isdir(full_path) else None

File: src/project-store.ts
import fs from "node:fs";
import path from "node:path";
import type { ProjectInfo } from "./types.js";
import { encodeProjectPath, getProjectsDir } from "./path-utils.js";
import { getSession, isAlive, listSessions } from "./session-store.js";
import { extractCwdFromJsonl } from "./jsonl-reader.js";

export function listProjects(claudeHome = "~/.claude"): ProjectInfo[] {
  const projectsDir = getProjectsDir(claudeHome);
  if (!fs.existsSync(projectsDir)) return [];

  const sessions = listSessions(claudeHome);
  const cwdToSessions = new Map<string, typeof sessions>();

  for (const s of sessions) {
    const list = cwdToSessions.get(s.cwd) || [];
    list.push(s);
    cwdToSessions.set(s.cwd, list);
  }

  const projects: ProjectInfo[] = [];
  for (const entry of fs.readdirSync(projectsDir).sort()) {
    const fullPath = path.join(projectsDir, entry);
    if (!fs.statSync(fullPath).isDirectory()) continue;

    // Resolve real path from session metadata (never decode directory name — lossy)
    const realPath = resolveProjectCwd(entry, fullPath, sessions);
    if (!realPath) continue;

    const jsonlFiles = fs.readdirSync(fullPath).filter((f) => f.endsWith(".jsonl"));
    const projectSessions = cwdToSessions.get(realPath) || [];

    projects.push({
      path: realPath,
      encoded_name: entry,
      session_count: jsonlFiles.length,
      has_active_session: projectSessions.some((s) => isAlive(s.pid)),
    });
  }

  return projects;
}

function resolveProjectCwd(
  encodedName: string,
  projectDir: string,
  sessions: { session_id: string; cwd: string }[],
): string | undefined {
  // Source 1: match encoded name against known session cwds
  for (const s of sessions) {
    if (encodeProjectPath(s.cwd) === encodedName) return s.cwd;
  }

  // Source 2: extract cwd from JSONL content (no lossy decode)
  const jsonlFiles = fs.readdirSync(projectDir).filter((f) => f.endsWith(".jsonl"));
  for (const jf of jsonlFiles) {
    const cwd = extractCwdFromJsonl(path.join(projectDir, jf));
    if (cwd) return cwd;
  }

  return undefined;
}

export function getProjectDir(projectPath: string, claudeHome = "~/.claude"): string | undefined {
  const encoded = encodeProjectPath(projectPath);
  const fullPath = path.join(getProjectsDir(claudeHome), encoded);
  return fs.existsSync(fullPath) && fs.statSync(fullPath).isDirectory() ? fullPath : undefined;
}
